newton/solvers/mujoco/solver_mujoco.py

```diff
@@ -1469,6 +1524,65 @@ def color_collision_shapes(model: Model, selected_shapes: nparray, visualize_gra
             shape_color = np.zeros(model.shape_count, dtype=np.int32)
         return shape_color
 
+    def create_newton_contact_geom_mapping(self, model: Model):
+        indices = [idx for w, idx in self.shape_map.values() if idx is not None]
+        if not indices:
+            return wp.array([], dtype=wp.int32, device=model.device)
+        max_mj_shape_id = max(indices)
+        geom_mapping = np.full((model.num_envs, max_mj_shape_id + 1), -1, dtype=np.int32)
+
+        for shape, (worldid, mj_geom) in self.shape_map.items():
+            if mj_geom is None:
+                continue
+            if worldid == -1:
+                worldid = slice(None)  # noqa
+            geom_mapping[worldid, mj_geom] = shape
+
+        return wp.array(geom_mapping, dtype=wp.int32, device=model.device)
+
+    def update_newton_contacts(self, model: Model, mj_data: MjWarpData, contact_info: ContactInfo):
+        nconmax = mj_data.nconmax
+
+        mj_contact = mj_data.contact
+
+        contact_info.ncon = mj_data.ncon
+        contact_info.position = mj_contact.pos
+        contact_info.separation = mj_contact.dist
+
+        if contact_info.pair is None:
+            contact_info.pair = wp.empty(nconmax, dtype=wp.vec2i, device=model.device)
+
+        if contact_info.normal is None:
+            contact_info.normal = wp.empty(nconmax, dtype=wp.vec3f, device=model.device)
+
+        if contact_info.force is None:
+            contact_info.force = wp.empty(nconmax, dtype=wp.float32, device=model.device)
+
+        num_threads = min(mj_data.nconmax, 8192)
```

Comment:
Does 8192 have to be the same number as `newton.utils.contact_sensor.NUM_THREADS`? Then you should probably use that.